src/ethereum/istanbul/vm/instructions/storage.py

```diff
@@ -0,0 +1,124 @@
+"""
+Ethereum Virtual Machine (EVM) Storage Instructions
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+.. contents:: Table of Contents
+    :backlinks: none
+    :local:
+
+Introduction
+------------
+
+Implementations of the EVM storage related instructions.
+"""
+from ethereum.base_types import U256
+from ethereum.utils.ensure import ensure
+
+from ...state import get_storage, set_storage
+from ...trie import trie_get
+from .. import Evm
+from ..error import OutOfGasError, WriteInStaticContext
+from ..gas import (
+    GAS_CALL_STIPEND,
+    GAS_SLOAD,
+    GAS_STORAGE_CLEAR_REFUND,
+    GAS_STORAGE_SET,
+    GAS_STORAGE_UPDATE,
+    subtract_gas,
+)
+from ..stack import pop, push
+
+
+def sload(evm: Evm) -> None:
+    """
+    Loads to the stack, the value corresponding to a certain key from the
+    storage of the current account.
+
+    Parameters
+    ----------
+    evm :
+        The current EVM frame.
+
+    Raises
+    ------
+    :py:class:`~ethereum.istanbul.vm.error.StackUnderflowError`
+        If `len(stack)` is less than `1`.
+    :py:class:`~ethereum.istanbul.vm.error.OutOfGasError`
+        If `evm.gas_left` is less than `50`.
+    """
+    evm.gas_left = subtract_gas(evm.gas_left, GAS_SLOAD)
+
+    key = pop(evm.stack).to_be_bytes32()
+    value = get_storage(evm.env.state, evm.message.current_target, key)
+
+    push(evm.stack, value)
+
+    evm.pc += 1
+
+
+def sstore(evm: Evm) -> None:
+    """
+    Stores a value at a certain key in the current context's storage.
+
+    Parameters
+    ----------
+    evm :
+        The current EVM frame.
+
+    Raises
+    ------
+    :py:class:`~ethereum.istanbul.vm.error.StackUnderflowError`
+        If `len(stack)` is less than `2`.
+    :py:class:`~ethereum.istanbul.vm.error.OutOfGasError`
+        If `evm.gas_left` is less than `20000`.
+    """
+    ensure(evm.gas_left > GAS_CALL_STIPEND, OutOfGasError)
+    ensure(not evm.message.is_static, WriteInStaticContext)
+
+    key = pop(evm.stack).to_be_bytes32()
+    new_value = pop(evm.stack)
+    current_value = get_storage(evm.env.state, evm.message.current_target, key)
+
+    _, original_trie = evm.env.state._snapshots[0]
```

Comment:
Reading internal details of `State` isn't compatible with the optimized `State`. This should be refactored into a function in `state.py` with signature:
```python
get_storage_original(state: State, address: Address, key: Bytes) -> U256
```

(I have no strong opinions on the name)

I will actually implement this function in the optimized state in a later PR.

Reply:
Updated